### DISCUSSION TITLE
Using multivariate forests for conditional prediction?

### DISCUSSION THREAD

**Discussion by smithja16** [kogalur/randomForestSRC#442] [Q&A]
Thank you for creating and maintaining randomForestSRC. I’m currently testing it for modelling and predicting multi-species communities, and I have a question about whether randomForestSRC is suitable for **conditional prediction**.

I use 'conditional prediction' to mean when we fit a multivariate model with all species, **then use this model to predict a subset of those species for new outcomes**. This unknown species subset is informed by any environmental predictors we included in the model as well as the abundance of the known subset of co-occurring species. I have used joint species modelling (the R packaged Hmsc) to do this, and it uses residual correlations among species estimated in the full model to then help inform the unknown species from the known species in the prediction. More info on conditional prediction can be found in: Wilkinson et al (2021). Defining and evaluating predictions of joint species distribution models. Methods in Ecology and Evolution, 12(3), 394-404.

I currently compare the Hmsc joint model against a set of **univariate random forests**. The univariate forests can create similar predictions simply by adding the ‘known’ species as predictor variables. But I’m wondering whether multivariate random forests can do something similar? I'm wondering if **imputing missing values is a kind of conditional prediction**. Does ‘impute’ in randomForestSRC use information from known species to inform species which have ‘NA’s?

Consider the following example data:
```
generate_biomass <- function(n, max_biomass, zero_prob) {
  biomass <- runif(n, min = 0, max = max_biomass)
  zeros <- rbinom(n, 1, zero_prob);   biomass * (1 - zeros) }

data <- data.frame(latitude = runif(15, min = -40, max = -30),
  depth = runif(15, min = 10, max = 50), sst = runif(15, min = 15, max = 25),
  spp1 = generate_biomass(15, 100, 0.2), spp2 = generate_biomass(15, 80, 0.3),
  spp3 = generate_biomass(15, 60, 0.4), spp4 = generate_biomass(15, 40, 0.5),
  spp5 = generate_biomass(15, 20, 0.6), spp6 = generate_biomass(15, 10, 0.7))

data[10:15,c("spp1","spp2","spp3")] <- NA
```
The final 5 rows for spp1, spp2, and spp3 are assumed unknown, and conditional prediction would predict those 15 values using their environmental responses **as well as their correlation with the remaining 3 species**. Normally I would first fit a model on training data with all species known then predict a species subset in new test data. 

However, I can generate these missing 'NA' values using impute, and **I've tried the following 3 impute options**, each of which give similar predictive performance:
```
formula <- as.formula("Multivar(spp1, spp2, spp3, spp4, spp5, spp6) ~ latitude + depth + sst")  #my full model has 46 species

imp1 <- impute(formula, data=data,  #my full data has 450 rows
              ntree=1000, nodesize=5, nimpute=5)

imp2 <- impute(formula, data=data,
              ntree=1000, nodesize=5, nimpute=5,
              splitrule="mahalanobis")

imp3 <- impute(formula, data=data,
              ntree=1000, nodesize=5, mf.q=0.5, max.iter=10) 
```
**My questions are these:**
1) whether the imputation I’m doing here is similar to ‘conditional prediction’ [are the biomasses of known species being used to help predict the unknown species in each ‘new’ outcome?]
2) which of the impute versions I’ve used best represents conditional prediction [or is there a better way, maybe using 'predict' on a trained model?]. 
3) does this method have any advantage over using a univariate forest for each unknown species (fitted to training data with the 3 known species as predictor variables, then predicted for the 'new' outcomes?)
 
I do note that when I make all species NA for the 'new' _n_ rows, the predictive performance is only a little worse, so perhaps the information provided by the biomass of my co-occurring species is not very valuable.

Thanks for any advice!

**Comment by ishwaran** [comment]
I would suggest using the missForest imputation method available in the `impute` function. The missForest imputation method has an interpretation as a prediction based method.  The method cycles through each variable treating it as an outcome to build a prediction model by regressing the non-missing outcome values on the other variables in the data.  This model is then used to estimate the missing outcome values by prediction.  A multivariate version of this can be invoked by appropriately setting the `mf.q` option which will have the interpretation of conditional prediction.

The code for implementing this on your example would look something like:

```
## standard missForest call
imp.c1 <- impute(data=data, mf.q=1)

## multivariate missForest call
imp.c2 <- impute(data=data, mf.q=3)
```

In the first call, each response is fit univariately, which means it conditions on all other species (and features) when building the prediction model.  Therefore this is a univariate regression where ssp1 is regressed against everything, ssp2 is regressed against everything and finally ssp3 is regressed against everything in a cyclical fashion.

The second call groups the responses into blocks of 3, which by the nature of your data since you have only 3 columns with missing data, will invoke multivariate regression using ssp1-ssp3 as the outcomes and regressing them on the remaining values.
I think this is the call you want.

PS:  there is a useful helper function `get.mv.formula` for building multivariate formula.  Here's a toy illustration.

```
> get.mv.formula(colnames(mtcars))
Multivar(mpg, cyl, disp, hp, drat, wt, qsec, vs, am, gear, carb) ~ 

```

**Reply by smithja16** [reply]
Thanks for this. On testing it seems about as accurate as the univariate RFs with other species as predictors.
